### ISSUE TITLE
Runtime apex overriding not working on Pixel 4 (uses a .apex zip)

### ISSUE THREAD

**Issue by Paschalis** [topjohnwu/Magisk#3008]
# Problem: I want to replace some bins/libs in the runtime apex
On a Google Pixel (_sailfish_/1st gen) I can successfully override the following runtime apex files with this module:

**mymodule:**
```
├── customize.sh
├── module.prop
└── system
    └── apex
        └── com.android.runtime.release
            ├── bin
            │   └── dex2oat
            └── lib64
                ├── libart-compiler.so
                └── libart.so
```

customize.sh:
```
# simply fixes permissions..
set_perm_recursive  $MODPATH                0       0       0755        0644
set_perm  $MODPATH/system/apex/com.android.runtime.release/bin/dex2oat 0       shell       0755
```

Installing the same module on a Google Pixel 4 (_flame_) the binaries/libs are not replacing the original ones.
The difference (and **probably the issue**) is that _Sailfish_ uses a folder for runtime apex:
`/system/apex/com.android.runtime.release/`
whereas _flame_ uses a `.apex` zip file:
`/system/apex/com.android.runtime.release.apex`

(also explained in another [issue comment](https://github.com/topjohnwu/Magisk/issues/2109#issuecomment-567303279))

Looking a bit on the code I can see [here](https://github.com/topjohnwu/Magisk/blob/master/scripts/util_functions.sh#L309) that there is support when the apex is a file, and it is supposed to unzip it, mount it, and then I guess replace with whatever I've put on my module in this path:
mymodule: `system/apex/com.android.runtime.release/`

Also on issue #2114  there was some preliminary support, but either part of the code was restructured/rewritten or wasn't included (I don't know the internals of `magisk`, so not very certain on this).
For example [here](https://github.com/topjohnwu/Magisk/blob/2c885344ddbe3c2f4e9cf5957869240e85234b44/scripts/util_functions.sh#L120) there's additional work on `recovery_actions`.

So is this a known issue?
Has anyone successfully replaced runtime apex bins/libs in cases where a zip file (.apex) file is used?
 
Thanks in advance!

# Additional info
Phone: Google Pixel 4/flame (The same module on Google Pixel/sailfish worked just fine)

Magisk Manager: 7.5.1
Magisk: 20.4.1

**Comment by topjohnwu** [comment]
Github issues are for reporting issues of Magisk itself, not for support on how to use Magisk, including developing modules.
The issue here might be the system verifying the signature of APEX zips

**Comment by Paschalis** [comment]
@topjohnwu how come this is not an issue, if a module that overrides runtime apex components works just fine on one device but on another does not?
(please see [comment here](https://github.com/topjohnwu/Magisk/issues/2109#issuecomment-567303279))

There isn't any mention in the documentation for this matter, not even in the [internal details](https://topjohnwu.github.io/Magisk/details.html), so where should I seek support?
Is there a dedicated forum for discussions around an involved issue like this one?
(Which is why I've shared the structure of my module, so anyone would know whether I've misconfigured something)

It's not clear for me whether this is properly supported by magisk or not. Can you please clarify this?

And regarding the signature verification, I haven't modified the apex zip or supplied a new one, instead I've supplied the components that I'd like to replace the original ones with, in the `/apex/com.android.runtime`.

**Comment by topjohnwu** [comment]
Oh after thinking about it, APEX modules are mounted before Magisk is ran. It is literally impossible to replace unless we push it pre-init, which I currently have no plans to support modules at that stage.

**Comment by topjohnwu** [comment]
It might be easier to directly replace the files in the apex folder after it is mounted. That might be significantly simpler.

**Comment by Paschalis** [comment]
Thanks for your reply. You get a lot spam in issues, so thanks for your time!

To replace after the apex folder is mounted, it means that my module should have something like:
```
apex
└── com.android.runtime
    ├── bin
    │   └── dex2oat
    └── lib64
        ├── libart-compiler.so
        └── libart.so
```
This is because the folder `/system/apex/com.android.runtime.release` gets mounted and renamed to:
`/apex/com.android.runtime`. It doesn't work however.
Is there a way to achieve this by using `customize.sh`?

I digged a little bit into the code and I found out that there is a bug in [mount_apex](https://github.com/topjohnwu/Magisk/blob/master/scripts/util_functions.sh#L301).
Initially, `/apex/` has nothing mounted to it.
When there is an apex zip file (line [309](https://github.com/topjohnwu/Magisk/blob/master/scripts/util_functions.sh#L309)), it unzips `apex_payload.img` and is put in `/apex`. However `loop_setup` fails to set anything in `LOOPDEV` as it is trying on current directory. So mount on line 315 never runs.

If the correct file path is set at lines 312, and 319 (`/apex/apex_payload.img`),
then magisk can successfully mount runtime (and the rest of the apex zip files) to `/apex/`.
Just like what android would do. **But when magisk is doing this?**
And what else should I change for this to pick my own changes that I've put at:
`/system/apex/com.android.runtime.release/' ?
I really need this functionality, and I could have the change to contribute a little bit to your awesome project.
But I need a bit of a guidance.

Should I instead extract `/system/apex/com.android.runtime.release.apex` in a temporary folder, say:
`/system/apex/com.android.runtime.release_/`, then merge it with whatever there is in the module at:
`/system/apex/com.android.runtime.release/`, and finally mount it at `/apex/com.android.runtime/`?

Thanks again!
